hpx/lcos/detail/future_data.hpp

```diff
@@ -388,6 +392,7 @@ namespace detail
 
             // check whether the data has already been set
             if (is_ready_locked()) {
+                l.unlock();
```

Comment:
This code can only be executed on a precondition violation, which would point to bugs elsewhere.